File: New Generation SmartPill/SmartPill/Model/CoreDataModels/Categories/Medicine+create.m
//
//  Remedios+create.m
//  SmartPill
//
//  Created by Dennis da Silva Nunes on 18/12/14.
//  Copyright (c) 2014 IFRN - Mobile School. All rights reserved.
//

#import "Medicine+create.h"



@implementation Medicine (create)

+ (Medicine *)medicineWithName:(NSString *)name
                  availability:(NSString *)availability
                        manufactuary:(NSString *)manufactuary
               activeIngredient:(NSString *)activeIngredient
                      quantity:(NSNumber *)quantity
                     reminder:(NSSet *)reminder
                          user:(User *)user
        inManagedObjectContext:(NSManagedObjectContext *)context{
    Medicine * medicine = nil;
    
    if ([name length]>0) {
     
        NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Medicine"];
        request.predicate = [NSPredicate predicateWithFormat:@"name = %@", name];
        
        NSError *error;
        NSArray *matches = [context executeFetchRequest:request error:&error];
        
        if (!matches) {
          
            return nil;
        } else if (![matches count]) {
           
            medicine = [NSEntityDescription insertNewObjectForEntityForName:@"Medicine"
                                                 inManagedObjectContext:context];
           
            medicine.name = name;
            medicine.availability = availability;
            medicine.manufacturer = manufactuary;
            medicine.activeIngredient = activeIngredient;
            medicine.quantity = quantity;
            medicine.reminder = reminder;
            [medicine addUserObject:user];
        } else {
            for (Medicine * med in matches) {
                if ([med.name isEqualToString:name]) {
                    return med;
                }
            }
        }
    }
    
    return medicine;
}

+ (void)updateMedicine:(Medicine *)medicine fromDataBaseContext:(NSManagedObjectContext*)context{
   
    if ([medicine.name length]>0) {
        
        NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Medicine"];
        request.predicate = [NSPredicate predicateWithFormat:@"name = %@", medicine.name];
        
        NSError *error;
        NSArray *matches = [context executeFetchRequest:request error:&error];
        
        if (!matches) {
            NSLog(@"Error in Update Medicine");
            return;
        } else if ([matches count]) {
            for (Medicine * med in matches) {
                if ([med.name isEqualToString:medicine.name]) {
                    med.name = medicine.name;
                    med.availability = medicine.availability;
                    med.manufacturer =medicine.manufacturer;
                    med.activeIngredient = medicine.activeIngredient;
                    med.quantity = medicine.quantity;
                    med.reminder = medicine.reminder;
                }
            }
        }else if(![matches count]){
            NSLog(@"Could not find any medicine in the database");
        
        }
        
    }
    [context save:nil];
}

@end

File: SmartPill/dbModals/Categories/Medicine+Create.swift
//
//  Medicine+Create.swift
//  SmartPill
//
//  Created by Dennis da Silva Nunes on 20/05/15.
//  Copyright (c) 2015 Dennis da Silva Nunes. All rights reserved.
//

import UIKit
import CoreData

extension Medicine{
    
    class func createMedicine(activeIngredient:String,availability:String,manufacturer:String,name:String,quantity:NSNumber,recipe:Recipe?,reminder:NSSet?)->Medicine?{
        
        let appDelegate = UIApplication.sharedApplication().delegate as! AppDelegate
        let context = appDelegate.managedObjectContext
        
        if(count(name)>0){
            let request = NSFetchRequest(entityName: "Medicine")
            request.predicate = NSPredicate(format: "name = %@", name)
            
            var matches = context!.executeFetchRequest(request, error: nil) as? [Medicine]
            
            if(matches == nil || matches!.count > 1){
                return nil
            }else if(matches!.count==0){
                //CREATE A NEW OFFER
                var med = NSEntityDescription.insertNewObjectForEntityForName("Medicine", inManagedObjectContext: context!) as! Medicine
                
                med.activeIngredient = activeIngredient
                med.availability = availability
                med.manufacturer = manufacturer
                med.name = name
                med.quantity = quantity
                if(recipe != nil){med.recipe = recipe!}
                if(reminder != nil){med.reminder = reminder!}
                context?.save(nil)
                return med
            }else{
                //OVERWRITE THE OLD OFFER
                var med:Medicine = matches!.first!
                
                med.activeIngredient = activeIngredient
                med.availability = availability
                med.manufacturer = manufacturer
                med.name = name
                med.quantity = quantity
                if(recipe != nil){med.recipe = recipe!}
                if(reminder != nil){med.reminder = reminder!}
                context?.save(nil)
                return med
            }
        }
        return nil
    }
}